### PR TITLE
Library

### PR DESCRIPTION
**Pull request by noklen**
Very very very big pull request))))

### CODE REVIEW

File: src/app/common/autorization-modal/autorization-modal.component.html [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<div class="session">
+  <div class="left"></div>
+  <form action="" class="log-in" autocomplete="off">
+    <!-- Display all of the SubContents is disable. -->
+    <div *ngIf="!signUp && !forgotPass">
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/autorization-modal.component.html]: Use only one flag example: modalType available values 'signIn', 'signUp', 'forgot'
modalType: enum

File: src/app/common/autorization-modal/autorization-modal.component.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  public isLogin = true;
+
+  constructor() { }
+
+  email = new FormControl('', [Validators.required, Validators.email]);
+
+  ngOnInit() {
+  }
+
+  getErrorMessage() {
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/autorization-modal.component.ts]: Add access modificator

File: src/app/common/autorization-modal/autorization-modal.component.html [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    <div *ngIf="signUp && !forgotPass">
+      <sign-up></sign-up>
+      <button mat-raised-button class="login-button switch-button"  (click)="signUp=!signUp">{{ 'signIn.button' | translate }}</button>
+    </div>
+    <div *ngIf="forgotPass && !signUp">
+      <reset-password></reset-password>
+      <button mat-raised-button class="login-button switch-button"  (click)="forgotPass=!forgotPass">{{ 'signIn.button' | translate }}</button>
+    </div>
+  </form>
+</div>
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/autorization-modal.component.html]: Use only one btn <button mat-raised-button class="login-button switch-button"  (click)="forgotPass=!forgotPass">{{ 'signIn.button' | translate }}</button>
remove duplicates

File: src/app/common/autorization-modal/autorization-modal.component.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  templateUrl: './autorization-modal.component.html',
+  styleUrls: ['./autorization-modal.component.scss'],
+})
+export class AutorizationModalComponent implements OnInit {
+
+  public isLogin = true;
+
+  constructor() { }
+
+  email = new FormControl('', [Validators.required, Validators.email]);
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/autorization-modal.component.ts]: Is it useble on html?

File: src/app/common/autorization-modal/autorization-modal.component.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+  ngOnInit() {
+  }
+
+  getErrorMessage() {
+    return this.email.hasError('required') ? 'You must enter a value' :
+      this.email.hasError('email') ? 'Not a valid email' :
+        '';
+  }
+
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/autorization-modal.component.ts]: Is it useble on html?

File: src/app/common/autorization-modal/log-in/log-in.component.html [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+      <mat-error *ngIf="email.invalid">{{getErrorMessage()}}</mat-error>
+    </mat-form-field>
+    <mat-form-field>
+      <input class="formInput" matInput placeholder="Enter your password" [type]="hide ? 'password' : 'text'" required>
+      <button mat-icon-button matSuffix (click)="hide = !hide" [attr.aria-label]="'Hide password'" [attr.aria-pressed]="hide">
+        <mat-icon aria-hidden="false">{{hide ? 'visibility_off' : 'visibility'}}</mat-icon>
+      </button>
+    </mat-form-field>
+  </div>
+    <button mat-raised-button class="login-button">{{ 'signIn.button' | translate }}</button>
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/log-in/log-in.component.html]: login or signIn. use one name

File: src/app/common/autorization-modal/log-in/log-in.component.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+@Component({
+  selector: 'log-in',
+  templateUrl: './log-in.component.html',
+  styleUrls: ['./log-in.component.scss']
+})
+export class LogInComponent implements OnInit {
+  hide = true;
+
+  email = new FormControl('', [Validators.required, Validators.email]);
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/log-in/log-in.component.ts]: acess modificators
add password
use reactiveForm

File: src/app/common/autorization-modal/log-in/log-in.component.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+  public getErrorMessage() {
+    return this.email.hasError('required') ? 'You must enter a value' :
+      this.email.hasError('email') ? 'Not a valid email' :
+        '';
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/log-in/log-in.component.ts]: acess modificators
Are you have validation on password?
use reactiveForm
move to html http://prntscr.com/r2bdry

File: src/app/common/autorization-modal/reset-password/reset-password.component.scss [hunk 9]
```diff
@@ -0,0 +1,5 @@
+.reset-button {
+  background:  rgb(182,157,230);
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/reset-password/reset-password.component.scss]: maybe move to hex format

File: src/app/common/autorization-modal/reset-password/reset-password.component.html [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+  <h4>We are <span>Help emigration</span></h4>
+  <p>Welcome back! In this page you can reset your password</p>
+  <div class="login-fields">
+    <mat-form-field>
+      <input class="formInput" matInput placeholder="Enter your email" [formControl]="email" required>
+      <mat-error *ngIf="email.invalid">{{getErrorMessage()}}</mat-error>
+    </mat-form-field>
+  </div>
+  <button mat-raised-button class="reset-button">{{ 'forgotPassword.reset' | translate }}</button>
+</div>
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/reset-password/reset-password.component.html]: Maybe added confirm password

File: src/app/common/autorization-modal/sign-up/sign-up.component.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    const pass = this.password.value;
+    const email = this.email.value;
+    const name = this.name.value;
+    console.log(email, pass);
+    this.dataService.registrationUser(email, pass, name).subscribe(data => {
+      console.log(data);
+    });
+  }
+
+}
```
Comment (AlexanderBeichuk) [inline src/app/common/autorization-modal/sign-up/sign-up.component.ts]: all for login

File: src/app/common/footer/footer.component.html [hunk 12]
```diff
@@ -1,21 +1,21 @@
...
+<!--    <div>-->
+<!--      <h5>LINKS</h5>-->
+<!--      <ul class="ul-footer">-->
+<!--        <li class="li-footer"><a>О проекте</a></li>-->
+<!--        <li class="li-footer"><a>Отзывы</a></li>-->
+<!--        <li class="li-footer"><a>Контакты</a></li>-->
+<!--        <li class="li-footer">Подписка</li>-->
+<!--      </ul>-->
+<!--    </div>-->
+<!--  </div>-->
```
Comment (AlexanderBeichuk) [inline src/app/common/footer/footer.component.html]: Remove if you dont use

File: src/app/common/header/header.component.ts [hunk 13]
```diff
@@ -12,8 +14,12 @@ export class HeaderComponent implements OnInit {
 
   public languages: Language[] = Object.keys(Language).map(key => Language[key]);
   public defaultLanguage: Language = this.languageService.getDefaultLanguage();
-
-  constructor(private languageService: LanguageService) {
+  public mainPage: boolean = this.currentPage();
```
Comment (AlexanderBeichuk) [inline src/app/common/header/header.component.ts]: its dont usable

File: src/app/common/header/header.component.ts [hunk 14]
```diff
@@ -22,4 +28,16 @@ export class HeaderComponent implements OnInit {
...
+  }
+
+  public currentPage (): boolean {
+    if (window.location.pathname !== '/home') {
+      console.log('123');
+      return false;
+    }
+    console.log('test');
+    return true;
+  }
```
Comment (AlexanderBeichuk) [inline src/app/common/header/header.component.ts]: its dont usable

File: src/app/models/url.enum.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+  CreateUser = 'user',
+  UpdateUser = 'update/user',
+  DeleteUser = 'delete/user/:id',*/
+  Registration = 'registration',
+  Login = 'login',
+  GetMe = 'me',
+  GetComments = 'comments',
+  CreateComment = 'comment',
+  UpdateComment = 'update/comment',
+  DeleteComment = 'delete/comment/:id',
```
Comment (AlexanderBeichuk) [inline src/app/models/url.enum.ts]: remove id

File: src/app/models/url.enum.ts [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+  GetFeedbacks = 'feedbacks',
+  CreateFeedback = 'feedback',
+  UpdateFeedback = 'update/feedback',
+  DeleteFeedback = 'delete/feedback/:id'
```
Comment (AlexanderBeichuk) [inline src/app/models/url.enum.ts]: remove id

File: src/app/common/main-menu/main-menu.component.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+  styleUrls: ['./main-menu.component.scss']
+})
+export class MainMenuComponent implements OnInit {
+
+  constructor() { }
+
+  ngOnInit() {
+  }
+
+}
```
Comment (AlexanderBeichuk) [inline src/app/common/main-menu/main-menu.component.ts]: Create new Module 'LibraryModule' and move all components for library to this module

File: src/app/pages/home-page/home-baner/home-baner.component.scss [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    left: 0;
+    opacity: 0.45;
+    position: absolute;
+    top: 0;
+    width: 100%;
+    z-index: 1;
+  }
+}
+
+@media screen and (max-width: 736px) {
```
Comment (AlexanderBeichuk) [inline src/app/pages/home-page/home-baner/home-baner.component.scss]: You have media in constants. Use constants

File: src/app/pages/home-page/home-baner/home-baner.component.scss [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+      margin-bottom: 0.5em;
+      padding-bottom: 0;
+    }
+    p {
+      font-size: 1.25em;
+    }
+  }
+}
+
+@media screen and (max-width: 480px) {
```
Comment (AlexanderBeichuk) [inline src/app/pages/home-page/home-baner/home-baner.component.scss]: You have media in constants. Use constants

File: src/app/pages/home-page/home-baner/home-baner.component.scss [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+  #banner p {
```
Comment (AlexanderBeichuk) [inline src/app/pages/home-page/home-baner/home-baner.component.scss]: use classes

File: src/app/pages/home-page/information/information.component.scss [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+  &:before {
+    -moz-transition: opacity 3s ease;
+    -webkit-transition: opacity 3s ease;
+    -ms-transition: opacity 3s ease;
+    transition: opacity 3s ease;
+    -moz-transition-delay: 1.25s;
+    -webkit-transition-delay: 1.25s;
+    -ms-transition-delay: 1.25s;
+    transition-delay: 1.25s;
+    content: '';
```
Comment (AlexanderBeichuk) [inline src/app/pages/home-page/information/information.component.scss]: content move to begin

File: src/app/pages/page-not-found/page-not-found.component.scss [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+          color: #262626;
+          margin: 0;
+          letter-spacing: 3px;
+          padding-left: 6px;
+        }
+      }
+    }
+  }
+
+  @media only screen and (max-width: 767px) {
```
Comment (AlexanderBeichuk) [inline src/app/pages/page-not-found/page-not-found.component.scss]: You have media in constants. Use constants

File: src/app/pages/page-not-found/page-not-found.component.scss [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+      .notfound-404 {
+        height: 200px;
+        h1 {
+          font-size: 200px;
+        }
+      }
+    }
+  }
+
+  @media only screen and (max-width: 480px) {
```
Comment (AlexanderBeichuk) [inline src/app/pages/page-not-found/page-not-found.component.scss]: You have media in constants. Use constants

File: src/app/services/data/data.service.ts [hunk 24]
```diff
@@ -0,0 +1,17 @@
+import { Injectable } from '@angular/core';
+import {HttpClient, HttpParams} from '@angular/common/http';
+import {Url} from '../../models/url.enum';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class DataService {
```
Comment (AlexanderBeichuk) [inline src/app/services/data/data.service.ts]: Rename to httpService

File: src/app/services/data/data.service.ts [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+  private REST_API_SERVER = 'http://localhost:3001/dev/api/';
```
Comment (AlexanderBeichuk) [inline src/app/services/data/data.service.ts]: Move to env

File: src/app/services/data/data.service.ts [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+
+  constructor(private httpClient: HttpClient) { }
+
+  public registrationUser (email: string, password: string, name: string) {
```
Comment (AlexanderBeichuk) [inline src/app/services/data/data.service.ts]: httpService should includes 'get' and 'post' methods
Create Module for authorization

File: src/app/common/autorization-modal/reset-password/reset-password.component.scss (hunk 9 above)
Comment (noklen) [reply src/app/common/autorization-modal/reset-password/reset-password.component.scss]: done

File: src/app/common/footer/footer.component.html (hunk 12 above)
Comment (noklen) [reply src/app/common/footer/footer.component.html]: done

File: src/app/common/header/header.component.ts (hunk 13 above)
Comment (noklen) [reply src/app/common/header/header.component.ts]: done

File: src/app/models/url.enum.ts (hunk 15 above)
Comment (noklen) [reply src/app/models/url.enum.ts]: done

File: src/app/models/url.enum.ts (hunk 16 above)
Comment (noklen) [reply src/app/models/url.enum.ts]: done

File: src/app/pages/home-page/home-baner/home-baner.component.scss (hunk 18 above)
Comment (noklen) [reply src/app/pages/home-page/home-baner/home-baner.component.scss]: done

File: src/app/services/data/data.service.ts (hunk 25 above)
Comment (noklen) [reply src/app/services/data/data.service.ts]: done

File: src/app/services/data/data.service.ts (hunk 24 above)
Comment (noklen) [reply src/app/services/data/data.service.ts]: done

File: src/app/services/data/data.service.ts (hunk 26 above)
Comment (noklen) [reply src/app/services/data/data.service.ts]: you fix it

File: src/app/pages/page-not-found/page-not-found.component.scss (hunk 23 above)
Comment (noklen) [reply src/app/pages/page-not-found/page-not-found.component.scss]: done

File: src/app/pages/page-not-found/page-not-found.component.scss (hunk 22 above)
Comment (noklen) [reply src/app/pages/page-not-found/page-not-found.component.scss]: done

File: src/app/pages/home-page/information/information.component.scss (hunk 21 above)
Comment (noklen) [reply src/app/pages/home-page/information/information.component.scss]: done

File: src/app/pages/home-page/home-baner/home-baner.component.scss (hunk 20 above)
Comment (noklen) [reply src/app/pages/home-page/home-baner/home-baner.component.scss]: done

File: src/app/pages/home-page/home-baner/home-baner.component.scss (hunk 19 above)
Comment (noklen) [reply src/app/pages/home-page/home-baner/home-baner.component.scss]: done

File: src/app/common/autorization-modal/autorization-modal.component.html (hunk 1 above)
Comment (noklen) [reply src/app/common/autorization-modal/autorization-modal.component.html]: done

File: src/app/common/autorization-modal/autorization-modal.component.html (hunk 3 above)
Comment (noklen) [reply src/app/common/autorization-modal/autorization-modal.component.html]: done

File: src/app/common/autorization-modal/log-in/log-in.component.html (hunk 6 above)
Comment (noklen) [reply src/app/common/autorization-modal/log-in/log-in.component.html]: done

File: src/app/common/autorization-modal/reset-password/reset-password.component.html (hunk 10 above)
Comment (noklen) [reply src/app/common/autorization-modal/reset-password/reset-password.component.html]: done

File: src/app/common/autorization-modal/autorization-modal.component.ts (hunk 5 above)
Comment (noklen) [reply src/app/common/autorization-modal/autorization-modal.component.ts]: done

File: src/app/common/autorization-modal/autorization-modal.component.ts (hunk 4 above)
Comment (noklen) [reply src/app/common/autorization-modal/autorization-modal.component.ts]: done

File: src/app/pages/home-page/home-baner/home-baner.component.scss (hunk 20 above)
Comment (AlexanderBeichuk) [reply src/app/pages/home-page/home-baner/home-baner.component.scss]: No

**Pull request merged**